lib/connection/client.rb

```diff
@@ -1,18 +1,87 @@
 module Connection
   class Client
     include Connection
-    include Connection::IOMethods
 
-    attr_reader :host
-    attr_reader :port
+    attr_reader :io
 
-    def initialize(host, port)
-      @host = host
-      @port = port
+    def initialize(io)
+      @io = io
     end
 
-    def socket
-      @socket ||= TCPSocket.new host, port
+    def self.build(io, scheduler=nil)
+      instance = new io
+      instance.configure_dependencies scheduler: scheduler
+      instance
+    end
+
+    def close
+      io.close
+      telemetry.closed
+
+    rescue IOError => error
+      telemetry.closed
+      raise error
+    end
+
+    def readline(*arguments)
+      readline_command.(*arguments)
+    end
+    alias_method :gets, :readline
+
+    def max_read_size
+      8192
```

Comment:
Not really. This is default amount we request when you call `read`. The reason for this is a little non obvious. The non-blocking variation of `IO#read`, `IO#read_nonblock` requires you to specify the number of bytes you'd like to read. This is due to the fact that it is based on lower level parts of the C standard library than `IO#read` (namely, `read` instead of `fread`).

This default was chosen to match the size of buffers that ruby uses for all IO objects.